In this project, I will be using Credit card approval prediction dataset from Specific sources

and the structure of the project will be follow:

* To get a basic introduction of the project and whats the business problem associate with it?
* start by loading and viewing the dataset.
* To manipulate the data, if there are any missing values in the dataset.
* To perform Exploratory data analysis(EDA) on the dataset
* To preproccessing the data before applying ML model to our dataset.
* Then, apply ML Model that can be predicted if an individual's application for credit card will be accepted or not.

#### Credit card applications and the problems associated with it

Nowadays, banks receives a lot of applications for issuance of credit cards. Many of them rejected for many reasons, like high-loan balances, low income levels, or too many inquiries on an individual's credit report. Manually analysing these applications is error prone and time consuming process, for this we can be automated with the power of ML and pretty much every bank does so nowadays. In this project, we will be build a automated credit card aproval predictor using ML techniques, just like real banks do.

### First Task

Importing the pandas package and loading the dataset

#####      1. Pandas : 
pandas is used to read the dataset file and import it as a dataframe, which is similar to a table with Rows and Columns

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
import os
import sys

In [4]:
os.getcwd()

'C:\\Users\\ar'

In [5]:
os.chdir("E:\\Data Sets")

In [6]:
# Loading the dataset from client repository or DB
# for that we have to use read_csv() method
cols = ["Gender", "Age", "Debt", "Married", "BankCustomer", "EducationLevel", "Ethnicity", "YearsEmployed", "PriorDefault", "Employed", "CreditScore", "DriversLicense", "Citizen", "ZipCode", "Income", "ApprovalStatus"]
df = pd.read_csv("credit-approval_csv.csv")
credit = df.rename(columns = {"A1":"Gender", "A2":"Age", "A3":"Debt", "A4":"Married", 'A5':"BankCustomer", 'A6':"EducationLevel", 'A7':"Ethnicity", 'A8':"YearsEmployed", 'A9':"PriorDefault", 'A10':"Employed", 'A11':"CreditScore", 'A12':"DriversLicense", 'A13':"Citizen", 'A14':"ZipCode", 'A15':"income", 'class':"ApprovalStatus"})
credit.head()                              

Gender    Age   Debt Married BankCustomer EducationLevel Ethnicity  \
0      b  30.83  0.000       u            g              w         v   
1      a  58.67  4.460       u            g              q         h   
2      a  24.50  0.500       u            g              q         h   
3      b  27.83  1.540       u            g              w         v   
4      b  20.17  5.625       u            g              w         v   

   YearsEmployed PriorDefault Employed  CreditScore DriversLicense Citizen  \
0           1.25            t        t            1              f       g   
1           3.04            t        t            6              f       g   
2           1.50            t        f            0              f       g   
3           3.75            t        t            5              t       g   
4           1.71            t        f            0              f       s   

   ZipCode  income ApprovalStatus  
0    202.0       0              +  
1     43.0     560              +  
2    280.0     824              +  
3    100.0       3              +  
4    120.0       0              +

on above, the output appears a bit confusing at its first sight, but lets try figure out the most important features of credit card application. we find that since data is confidential, the contributor of this dataste has anonymized the feature names. The features of this dataset have been anonymized to protect the privacy, but it gives us a pretty good overview of the probably features.

The probably features in a typical credit card applications are:

* Gender
* Age
* Debt
* Married
* BankCustomer
* EducationLevel
* Ethnicity
* YearsEmployed
* PriorDefault
* Employed
* CreditScore
* DriverseLicense
* Citizen
* ZipCode
* Income
* ApprovalStatus

### Second Task

The Dataset has a mixture of numerical and non-numerical features. It can be fixed with some preprocessing, but before we do that, lets learn about the dataset a bit more to see if there are other dataset issues that nedd to be fixed.

So, lets start by printing the summary statistics using describe() method and dataframe information using info() method

In [5]:
##### To print summary statistics
print(credit.describe())
print(credit.describe(include = ['O']))

print("The dataset give the 6 features are numeric and remainings are non numeric features.\nThe Describe() method is summary of statistics only for numeric features")

              Age        Debt  YearsEmployed  CreditScore      ZipCode  \
count  678.000000  690.000000     690.000000    690.00000   677.000000   
mean    31.568171    4.758725       2.223406      2.40000   184.014771   
std     11.957862    4.978163       3.346513      4.86294   173.806768   
min     13.750000    0.000000       0.000000      0.00000     0.000000   
25%     22.602500    1.000000       0.165000      0.00000    75.000000   
50%     28.460000    2.750000       1.000000      0.00000   160.000000   
75%     38.230000    7.207500       2.625000      3.00000   276.000000   
max     80.250000   28.000000      28.500000     67.00000  2000.000000   

              income  
count     690.000000  
mean     1017.385507  
std      5210.102598  
min         0.000000  
25%         0.000000  
50%         5.000000  
75%       395.500000  
max    100000.000000  
       Gender Married BankCustomer EducationLevel Ethnicity PriorDefault  \
count     678     684          684            681 

In [6]:
credit["DriversLicense"].value_counts()

f    374
t    316
Name: DriversLicense, dtype: int64

In [7]:
# To print dataset information
credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Gender          678 non-null    object 
 1   Age             678 non-null    float64
 2   Debt            690 non-null    float64
 3   Married         684 non-null    object 
 4   BankCustomer    684 non-null    object 
 5   EducationLevel  681 non-null    object 
 6   Ethnicity       681 non-null    object 
 7   YearsEmployed   690 non-null    float64
 8   PriorDefault    690 non-null    object 
 9   Employed        690 non-null    object 
 10  CreditScore     690 non-null    int64  
 11  DriversLicense  690 non-null    object 
 12  Citizen         690 non-null    object 
 13  ZipCode         677 non-null    float64
 14  income          690 non-null    int64  
 15  ApprovalStatus  690 non-null    object 
dtypes: float64(4), int64(2), object(10)
memory usage: 86.4+ KB


Here, our dataset contains both numerical and non-numerical data(specifically data that are of "float64", "int64" and "object"), specifically that features are *Age*, *Debt*, *YearsEmployed*, *CreditScore*, *ZipCode* and *Income* features contains numerical values(float64, int64) and remaining all features contains non-numerical types(object)

### Third Task
##### Manipulating the data
###### We have uncovered some issues that will effect the performance of our ML model if they go unchanged:

In [8]:
credit.iloc[671:681,:]

Gender    Age    Debt Married BankCustomer EducationLevel Ethnicity  \
671      b  25.83  12.835       u            g             cc         v   
672      a  50.25   0.835       u            g             aa         v   
673    NaN  29.50   2.000       y            p              e         h   
674      a  37.33   2.500       u            g              i         h   
675      a  41.58   1.040       u            g             aa         v   
676      a  30.58  10.665       u            g              q         h   
677      b  19.42   7.250       u            g              m         v   
678      a  17.92  10.210       u            g             ff        ff   
679      a  20.08   1.250       u            g              c         v   
680      b  19.50   0.290       u            g              k         v   

     YearsEmployed PriorDefault Employed  CreditScore DriversLicense Citizen  \
671          0.500            f        f            0              f       g   
672          0.500            f        f            0              t       g   
673          2.000            f        f            0              f       g   
674          0.210            f        f            0              f       g   
675          0.665            f        f            0              f       g   
676          0.085            f        t           12              t       g   
677          0.040            f        t            1              f       g   
678          0.000            f        f            0              f       g   
679          0.000            f        f            0              f       g   
680          0.290            f        f            0              f       g   

     ZipCode  income ApprovalStatus  
671      0.0       2              -  
672    240.0     117              -  
673    256.0      17              -  
674    260.0     246              -  
675    240.0     237              -  
676    129.0       3              -  
677    100.0       1              -  
678      0.0      50              -  
679      0.0       0              -  
680    280.0     364              -

In [9]:
credit.isnull().sum()

Gender            12
Age               12
Debt               0
Married            6
BankCustomer       6
EducationLevel     9
Ethnicity          9
YearsEmployed      0
PriorDefault       0
Employed           0
CreditScore        0
DriversLicense     0
Citizen            0
ZipCode           13
income             0
ApprovalStatus     0
dtype: int64

Dataset has missing values, which we will takecare of in the manipulating process, the missing values in the dataset are labeled with *NaN*.
we have to treat the missing values that we are going to perform manipulating process.

An important question that get raised here is  `Why are we giving so much importance to missing values`? can't they be just ignored? ignoring missing values can effect the performance of a ML model heavily. While ignoring the missing values our model may miss out on information about the dataset that may be useful for its training. Then there are many models which can't handle missing values implicitly such as LDA(Linear Descriminant Analysis).

So, to avoid this problem, we are going to impute the missing values with a stretegy called *Mean Imputation*.

In [25]:
# Impute the missing values with Mean imputation
credit.fillna(credit.mean(), inplace = True)

In [11]:
credit.isnull().sum()

Gender            12
Age                0
Debt               0
Married            6
BankCustomer       6
EducationLevel     9
Ethnicity          9
YearsEmployed      0
PriorDefault       0
Employed           0
CreditScore        0
DriversLicense     0
Citizen            0
ZipCode            0
income             0
ApprovalStatus     0
dtype: int64

We have successfully taken of the missing values present in the numeric columns. There are still some missing values to be imputed for columns like `Gender, Married, BankCustomer, EducationalLevel, Ethnicity`. All of these columns contains non-numerical data and this why the mean imputation stretagy would not work here. For this needs a different treatment.

We are going to impute these missing values with the most frequent values as present in the repsective columns. For this we have to write the for loop.


In [28]:
credit["Gender"].dtype == 'int64'

False

In [13]:
#credit["Married"].value_counts().index[0]

In [14]:
#credit.fillna(credit["Gender"].value_counts())

In [15]:
# Iterate over each column of dataset
for col in credit:
    # Check if the column is object type
    if credit[col].dtype == 'object':
        # Impute the most frequent value
        credit = credit.fillna(credit[col].value_counts().index[0])
        
credit.isna().sum()

Gender            0
Age               0
Debt              0
Married           0
BankCustomer      0
EducationLevel    0
Ethnicity         0
YearsEmployed     0
PriorDefault      0
Employed          0
CreditScore       0
DriversLicense    0
Citizen           0
ZipCode           0
income            0
ApprovalStatus    0
dtype: int64

In above, impute the missing values in non-numerical features with most frequent values for each column

* Missing values are successfully handled


### Fourth Task
#### Preprocessing the data

In the data, there are still minor but essential data preprocessing needed before we proceed towards building ML model.
We are going to divide these remaining preprocessing steps into three main tasks:

  1. Convert the non-numeric data into numeric.
  2. Split the data into train and test sets.
  3. Scale the feature values to a uniform range.

First, we will converting all the non-numeric into numeric ones. We do this because not only it results in a faster computation but also many ML Models (like XGBoost) require the data to be in a strictly numeric format. We will do this by using a techinque called $LabelEncoding$

In [16]:
# First, Import the label encoder
from sklearn.preprocessing import LabelEncoder

# Now, instantiated LabelEncoder
le = LabelEncoder()

# Iterate over all the values of each column and extract their dtypes
for col in credit:
    # compare if the dtype is 'object'
    if credit[col].dtype == 'object':
        # use labelEncoder to do the numeric_transformation
        credit[col] = le.fit_transform(credit[col])
        
print(credit.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Gender          690 non-null    int32  
 1   Age             690 non-null    float64
 2   Debt            690 non-null    float64
 3   Married         690 non-null    int32  
 4   BankCustomer    690 non-null    int32  
 5   EducationLevel  690 non-null    int32  
 6   Ethnicity       690 non-null    int32  
 7   YearsEmployed   690 non-null    float64
 8   PriorDefault    690 non-null    int32  
 9   Employed        690 non-null    int32  
 10  CreditScore     690 non-null    int64  
 11  DriversLicense  690 non-null    int32  
 12  Citizen         690 non-null    int32  
 13  ZipCode         690 non-null    float64
 14  income          690 non-null    int64  
 15  ApprovalStatus  690 non-null    int32  
dtypes: float64(4), int32(10), int64(2)
memory usage: 59.4 KB
None


In the dataset, some features are not important like `DriversLicense, ZipCode` features. So we should drop them to design our ML Model with the best set of features, This concept in *Data Science* is $Feature Selection$

In [17]:
# To drop specific features using drop() method
credit1 = credit.drop(['ZipCode','DriversLicense'], axis = 1)

In [18]:
credit1.columns

Index(['Gender', 'Age', 'Debt', 'Married', 'BankCustomer', 'EducationLevel',
       'Ethnicity', 'YearsEmployed', 'PriorDefault', 'Employed', 'CreditScore',
       'Citizen', 'income', 'ApprovalStatus'],
      dtype='object')

### Fifth Task
#### Spliting the data

Now, we split our dataset into train and test sets,to preprare our data for two different phases of ML modeling: Training and testing.

In [19]:
# Importing the train_test_split() method from sklearn.model_selection
from sklearn.model_selection import train_test_split

In [20]:
# First, Convert dataFrame into Numpy array and
# Segregate Features and labels into separate values
X, y = credit1.iloc[:, 0:13].values, credit1.iloc[:, -1].values
print([X.shape,y.shape])

[(690, 13), (690,)]


In [21]:
# Now, Split Train and Test sets
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.25, random_state = 1234)
print([X_train.shape, X_test.shape, y_train.shape, y_test.shape])

[(517, 13), (173, 13), (517,), (173,)]


### Sixth task
#### Feature Scalling

The dataset also contains values from several ranges. Some features have a range of *0-28*, some features have a range of *2-67* and some have a range of *1017-100000*. Apart from these, we can get useful statistical information like *mean*, *Min* and *Max* about the features that have numerical values.

Ideally, no information from the test data should be used to scale the training data or should be used to direct the training process of a machine learning model. 
Hence, we first split the data and then apply the scaling.
Here, we have only one final preprocessing step before applying the ML model to data.

Now, lets try to understand what the scaled values mean in the real world.
Here, lets use `CreditScore` feature as example, the credit score of a person is their creditworthiness based on their credit history.
Higher this number, the more financially trustworthy a person is consider to be.
So, `CreditScore` of $1$ is the Highest since we are rescalling all the values to the range of $0-1$.

For this we have to import $MinMaxScaler()$ method from *sklearn.preprocessing*

In [22]:
# Import MinMaxScaler() method
from sklearn.preprocessing import MinMaxScaler

# Now Instantiate MinMaxScaler with mms_scaler variable and give "feature_range" parameter value is 0-1
mms_scaler = MinMaxScaler(feature_range = (0,1))

# Use it to rescale X_train and X_test
rescaled_X_train = mms_scaler.fit_transform(X_train)
rescaled_X_test = mms_scaler.fit_transform(X_test)
print(rescaled_X_train)

[[1.         0.20932331 0.0297501  ... 0.         0.         0.00251   ]
 [1.         0.40977444 0.00158667 ... 0.04347826 1.         0.        ]
 [1.         0.09518797 0.0099167  ... 0.         0.         0.        ]
 ...
 [1.         0.33954887 0.08429195 ... 0.04347826 0.         0.01187   ]
 [1.         0.31834586 0.099167   ... 0.         0.         0.002     ]
 [0.         0.03503759 0.01328838 ... 0.04347826 0.         0.00126   ]]


### Seventh Task
#### Build the model to the training set

Essentially, predicting if a credit card application will be approved or not is a classification task.
In our dataset contains more instances that correpond to "Denied" status then instances corresponding to "Approved" status.
Specifically, out of 100% instances, there are 55.5% applications that got "Denied" and 44.5% applications got "Approved"

It gives us benchmark. A good ML model should be able to accuratly predict the status of the applications with respect to these statistics.

Here, question rises like *Which model should we pick?* a question to ask is: *are the features that effect the credit card approval decision process correlated with each other?*. Although we can measure correlation, because of this correlation we will take advantage of the fact that generalized linear models perform well in these cases.


##### sklearn package:
  This ML library includes numerous ML algorithms already built-in with certain parameters set as default parameters so they work right out of the box

In [23]:
# Now, Import Logistic Regression
from sklearn.linear_model import LogisticRegression

# Here, instantiate a Logisticregression classifier with defaul parameter values
lgr = LogisticRegression()

# Now, fit lgr model with train set
lgr.fit(rescaled_X_train, y_train)

LogisticRegression()

### Eight Task
#### Making the predictions and evaluating the performance of the model

Here, How well does our model perform?
We will now evaluate our model on the test set with respect to classification accuracy.
But, we will also take look the model's confusion matrix.
In the case of predicting credit card applications, it is equally important to see if our ML model is able to predict the approval status of the application as "Denied" that originally got "Denied". If our model is not perform well in this aspect, then it might end up approving the application that should have been approved. The confusion matrix help us to view our model performance from these aspects.

In [32]:
# Import Confusion_Matrix from sklearn.metrics
from sklearn.metrics import confusion_matrix

# Use lgr model to predict instances from the rescaled test set and store it into y_pred variable
y_pred = lgr.predict(rescaled_X_test)

#print(y_pred)
#print(y_test)
# Now, get the accuracy score of lgr model and print it
print("Accuracy of Logistic Regression classifier is:", lgr.score(rescaled_X_test, y_test))
print("\n")

# Now print the Confusion matrix of lgr model
print("The Confusion Matrix of model is:\n", confusion_matrix(y_test, y_pred))

Accuracy of Logistic Regression classifier is: 0.8439306358381503


The Confusion Matrix of model is:
 [[67 11]
 [16 79]]


### Ninth Task
#### GridSearch for model performance better

The model is pretty good! It was able to yield an accuracy score of almost 84.4%.

For the confusionMatrix, the first element of the first row of the confusion matrix denote the TrueNegative meaning the no.ofnegative instances (denied applications) predicted by the model correctly